### PR TITLE
Sensor struct

### PR DESCRIPTION
**Pull request by qwandor-google**
This introduces a struct to contain information about a sensor, and uses it to automatically try reconnecting if no readings are seen for a while.

### CODE REVIEW

File: publish-mqtt/src/main.rs [hunk 1]
```diff
@@ -107,37 +108,67 @@ async fn main() -> Result<(), Box<dyn Error + Send + Sync>> {
...
+        })
+    }
+
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    pub fn mac_address(&self) -> &str {
+        &self.mac_address
+    }
```
Comment (alsuren) [inline publish-mqtt/src/main.rs]: (controversial) I don't like getters. Can we just make the fields public and trust people to not fuck with them?

File: publish-mqtt/src/main.rs [hunk 2]
```diff
@@ -208,27 +249,34 @@ async fn requests(mut homie: HomieDevice) -> Result<(), Box<dyn Error>> {
...
-                        homie
-                            .publish_value(&node_id, "battery", readings.battery_percent)
-                            .await?;
+                    if let Some(sensor) = sensors_connected
+                        .iter_mut()
+                        .find(|s| s.device_path == device_path)
+                    {
+                        sensor.last_update_timestamp = Instant::now();
+                        if let Some(readings) = decode_value(&value) {
+                            println!("{} {} ({})", sensor.mac_address(), readings, sensor.name());
```
Comment (alsuren) [inline publish-mqtt/src/main.rs]:
```suggestion
                        if let Some(readings) = decode_value(&value) {
                            println!("{} {} ({})", sensor.mac_address(), readings, sensor.name());
                            sensor.last_update_timestamp = Instant::now();
```

If we get a reading that's bogus, does it really count as a reading?

File: publish-mqtt/src/main.rs [hunk 3]
```diff
@@ -167,35 +199,44 @@ async fn requests(mut homie: HomieDevice) -> Result<(), Box<dyn Error>> {
...
+            .position(|s| now - s.last_update_timestamp > UPDATE_TIMEOUT)
+        {
+            let sensor = sensors_connected.remove(sensor_index);
+            println!(
+                "No update from {} for {:?}, reconnecting",
+                sensor.name,
+                now - sensor.last_update_timestamp
+            );
+            homie.remove_node(&sensor.node_id()).await?;
+            sensors_to_connect.push_back(sensor);
```
Comment (alsuren) [inline publish-mqtt/src/main.rs:229]: Should we also disconnect from it via bluetooth? Would this increase our chances of getting properly connected when it comes up in the queue?

Comment (qwandor-google) [reply publish-mqtt/src/main.rs:229]: I'm not entirely sure. From when I was running it before it seemed to work without disconnecting.

File: publish-mqtt/src/main.rs (hunk 2 above)
Comment (qwandor-google) [reply publish-mqtt/src/main.rs]: As far as being fairly confident that we are still connected, I think it does count.

File: publish-mqtt/src/main.rs (hunk 1 above)
Comment (qwandor-google) [reply publish-mqtt/src/main.rs]: Done.

**Review by alsuren — APPROVED** [review]

**Pull request merged**